src/sequence.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Random operations on sequences
+
+use Rng;
+use dist::range;
+
+/// This trait implements a `choose` operations on slices and sequences.
+pub trait Choose<T> {
+    /// Return one element from a sequence.
+    /// 
+    /// Returns `None` only if the sequence is empty.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use rand::{thread_rng, Choose};
+    ///
+    /// let choices = [1, 2, 4, 8, 16, 32];
+    /// let mut rng = thread_rng();
+    /// println!("{:?}", choices[..].choose(&mut rng));
+    /// assert_eq!(choices[..0].choose(&mut rng), None);
+    /// ```
+    fn choose<R: Rng>(self, rng: &mut R) -> Option<T>;
```

Comment:
About dynamic dispatch: The new trait looks a lot better prepared for trait objects already. Wide use of `<R: ?Sized + Rng>` through the crate should make it possible to use dynamic dispatch, it's just a major breaking change of Rand and other trait implementors, but if that's not an issue, we can do it.

(There's two ways to go about it, using `R: ?Sized + Rng` with a `&mut R` parameter or using `R: Rng` and an `R` parameter, both work the same way if you have a blanket implementation of `Rng for &mut R where R: Rng`; I *think* the `&mut R` parameter is just as good a choice, but the second form could make it easier if you have Rng "adaptors" or whatever..)

Either way, better usability of Rng from trait objects is something I've wished for, but it's not relevant for all use cases.